dbms/src/Common/Elf.cpp

```diff
@@ -1,4 +1,4 @@
-#ifdef __ELF__
+#if defined(__ELF__) && !defined(__FreeBSD__)
```

Comment:
Why we don't want support ELF stuff on FreeBSD?

Reply:
Maybe we want, but it requires lot of investigation 17e93b8f2d9672616a2e5b15148258467886a423
```
1/314] Running C++ protocol buffer compiler on /root/ClickHouse/utils/test-data-generator/../../dbms/tests/queries/0_stateless/00825_protobuf_format.proto                                                                                                                                                                                                                                                                                                                                 
FAILED: utils/test-data-generator/00825_protobuf_format.pb.cc utils/test-data-generator/00825_protobuf_format.pb.h                                                                                                                                                                                                                                                                                                                                                                          
cd /root/ClickHouse/build_debug/utils/test-data-generator && /root/ClickHouse/build_debug/contrib/protobuf/cmake/protoc --cpp_out /root/ClickHouse/build_debug/utils/test-data-generator -I /root/ClickHouse/dbms/tests/queries/0_stateless /root/ClickHouse/dbms/tests/queries/0_stateless/00825_protobuf_format.proto                                                                                                                                                                     
/usr/local/lib/gcc7/libstdc++.so.6: version GLIBCXX_3.4.26 required by /root/ClickHouse/build_debug/contrib/protobuf/cmake/protoc not found                                                                                                                                                                                                                                                                                                                                                 
[2/314] Running C++ protocol buffer compiler on /root/ClickHouse/utils/test-data-generator/../../dbms/tests/queries/0_stateless/00825_protobuf_format_syntax2.proto                                                                                                                                                                                                                                                                                                                         
FAILED: utils/test-data-generator/00825_protobuf_format_syntax2.pb.cc utils/test-data-generator/00825_protobuf_format_syntax2.pb.h                                                                                                                                                                                                                                                                                                                                                          
cd /root/ClickHouse/build_debug/utils/test-data-generator && /root/ClickHouse/build_debug/contrib/protobuf/cmake/protoc --cpp_out /root/ClickHouse/build_debug/utils/test-data-generator -I /root/ClickHouse/dbms/tests/queries/0_stateless /root/ClickHouse/dbms/tests/queries/0_stateless/00825_protobuf_format_syntax2.proto                                                                                                                                                             
/usr/local/lib/gcc7/libstdc++.so.6: version GLIBCXX_3.4.26 required by /root/ClickHouse/build_debug/contrib/protobuf/cmake/protoc not found                                                                                                                                                                                                                                                                                                                                                 
[3/314] Building CXX object dbms/src/Common/tests/CMakeFiles/symbol_index.dir/symbol_index.cpp.o                                                                                                                                                                                                                                                                                                                                                                                            
FAILED: dbms/src/Common/tests/CMakeFiles/symbol_index.dir/symbol_index.cpp.o                                                                                                                                                                                                                                                                                                                                                                                                                
/usr/local/bin/ccache /usr/local/bin/g++9  -DBOOST_SYSTEM_NO_DEPRECATED -DCLOCK_MONOTONIC_COARSE=CLOCK_MONOTONIC_FAST -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -Idbms/src/Core/include -I../dbms/src -Idbms/src -I../libs/libcommon/include -Ilibs/libcommon/include -I../contrib/cityhash102/include -I../contrib/croaring -isystem ../contrib/double-conversion -isystem ../contrib/re2 -isystem ../contrib/libpcg-random/include -isystem ../contrib/poco/Util/include -isystem ../contrib/poc
In file included from ../dbms/src/Common/SymbolIndex.h:7,                                                                                                                                                                                                                                                                                                                                                                                                                                   
                 from ../dbms/src/Common/tests/symbol_index.cpp:1:                                                                                                                                                                                                                                                                                                                                                                                                                          
../dbms/src/Common/Elf.h:16:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   16 | using ElfAddr = ElfW(Addr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:17:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   17 | using ElfEhdr = ElfW(Ehdr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:18:16: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   18 | using ElfOff = ElfW(Off);                                                                                                                                                                                                                                                                                                                                                                                                                                                           
      |                ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
../dbms/src/Common/Elf.h:19:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   19 | using ElfPhdr = ElfW(Phdr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:20:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   20 | using ElfShdr = ElfW(Shdr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:21:16: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   21 | using ElfSym = ElfW(Sym);                                                                                                                                                                                                                                                                                                                                                                                                                                                           
      |                ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
../dbms/src/Common/Elf.h:34:15: error: 'ElfShdr' does not name a type; did you mean 'Elf_Shdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   34 |         const ElfShdr & header;                                                                                                                                                                                                                                                                                                                                                                                                                                                     
      |               ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                               
      |               Elf_Shdr                                                                                                                                                                                                                                                                                                                                                                                                                                                              
../dbms/src/Common/Elf.h:41:23: error: 'ElfShdr' does not name a type; did you mean 'Elf_Shdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   41 |         Section(const ElfShdr & header_, const Elf & elf_);                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                       ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                       
      |                       Elf_Shdr                                                                                                                                                                                                                                                                                                                                                                                                                                                      
../dbms/src/Common/Elf.h:61:11: error: 'ElfEhdr' does not name a type; did you mean 'Elf_Ehdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   61 |     const ElfEhdr * header;                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |           ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
      |           Elf_Ehdr                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
../dbms/src/Common/Elf.h:62:11: error: 'ElfShdr' does not name a type; did you mean 'Elf_Shdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   62 |     const ElfShdr * section_headers;                                                                                                                                                                                                                                                                                                                                                                                                                                                
      |           ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
      |           Elf_Shdr                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
[4/314] Building CXX object dbms/CMakeFiles/clickhouse_common_io.dir/src/Common/Elf.cpp.o                                                                                                                                                                                                                                                                                                                                                                                                   
FAILED: dbms/CMakeFiles/clickhouse_common_io.dir/src/Common/Elf.cpp.o                                                                                                                                                                                                                                                                                                                                                                                                                       
/usr/local/bin/ccache /usr/local/bin/g++9  -DBOOST_SYSTEM_NO_DEPRECATED -DCLOCK_MONOTONIC_COARSE=CLOCK_MONOTONIC_FAST -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -I../libs/libcommon/include -Ilibs/libcommon/include -Idbms/src/Core/include -I../dbms/src -Idbms/src -I../contrib/cityhash102/include -I../libs/libwidechar_width -I../contrib/croaring -isystem ../contrib/double-conversion -isystem ../contrib/brotli/c/include -isystem ../contrib/re2 -isystem ../contrib/libpcg-random/incl
In file included from ../dbms/src/Common/Elf.cpp:4:                                                                                                                                                                                                                                                                                                                                                                                                                                         
../dbms/src/Common/Elf.h:16:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   16 | using ElfAddr = ElfW(Addr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:17:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   17 | using ElfEhdr = ElfW(Ehdr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:18:16: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   18 | using ElfOff = ElfW(Off);                                                                                                                                                                                                                                                                                                                                                                                                                                                           
      |                ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
../dbms/src/Common/Elf.h:19:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   19 | using ElfPhdr = ElfW(Phdr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:20:17: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   20 | using ElfShdr = ElfW(Shdr);                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                 ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                
../dbms/src/Common/Elf.h:21:16: error: expected type-specifier before 'ElfW'                                                                                                                                                                                                                                                                                                                                                                                                                
   21 | using ElfSym = ElfW(Sym);                                                                                                                                                                                                                                                                                                                                                                                                                                                           
      |                ^~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                 
../dbms/src/Common/Elf.h:34:15: error: 'ElfShdr' does not name a type; did you mean 'Elf_Shdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   34 |         const ElfShdr & header;                                                                                                                                                                                                                                                                                                                                                                                                                                                     
      |               ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                               
      |               Elf_Shdr                                                                                                                                                                                                                                                                                                                                                                                                                                                              
../dbms/src/Common/Elf.h:41:23: error: 'ElfShdr' does not name a type; did you mean 'Elf_Shdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   41 |         Section(const ElfShdr & header_, const Elf & elf_);                                                                                                                                                                                                                                                                                                                                                                                                                         
      |                       ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                       
      |                       Elf_Shdr                                                                                                                                                                                                                                                                                                                                                                                                                                                      
../dbms/src/Common/Elf.h:61:11: error: 'ElfEhdr' does not name a type; did you mean 'Elf_Ehdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   61 |     const ElfEhdr * header;                                                                                                                                                                                                                                                                                                                                                                                                                                                         
      |           ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
      |           Elf_Ehdr                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
../dbms/src/Common/Elf.h:62:11: error: 'ElfShdr' does not name a type; did you mean 'Elf_Shdr'?                                                                                                                                                                                                                                                                                                                                                                                             
   62 |     const ElfShdr * section_headers;                                                                                                                                                                                                                                                                                                                                                                                                                                                
      |           ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                   
      |           Elf_Shdr                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
../dbms/src/Common/Elf.cpp: In constructor 'DB::Elf::Elf(const string&)':                                                                                                                                                                                                                                                                                                                                                                                                                   
../dbms/src/Common/Elf.cpp:24:27: error: 'ElfEhdr' was not declared in this scope; did you mean 'Elf_Ehdr'?                                                                                                                                                                                                                                                                                                                                                                                 
   24 |     if (elf_size < sizeof(ElfEhdr))                                                                                                                                                                                                                                                                                                                                                                                                                                                 
      |                           ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                   
      |                           Elf_Ehdr                                                                                                                                                                                                                                                                                                                                                                                                                                                  
../dbms/src/Common/Elf.cpp:28:5: error: 'header' was not declared in this scope                                                                                                                                                                                                                                                                                                                                                                                                             
   28 |     header = reinterpret_cast<const ElfEhdr *>(mapped);                                                                                                                                                                                                                                                                                                                                                                                                                             
      |     ^~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                                                          
../dbms/src/Common/Elf.cpp:28:37: error: ISO C++ forbids declaration of 'type name' with no type [-fpermissive]                                                                                                                                                                                                                                                                                                                                                                             
   28 |     header = reinterpret_cast<const ElfEhdr *>(mapped);                                                                                                                                                                                                                                                                                                                                                                                                                             
      |                                     ^~~~~~~                                                                                                                                                                                                                                                                                                                                                                                                                                         
../dbms/src/Common/Elf.cpp:28:37: error: expected '>' before 'ElfEhdr'                                                                                                                                                                                                                                                                                                                                                                                                                      
../dbms/src/Common/Elf.cpp:28:37: error: expected '(' before 'ElfEhdr'                                                                                                                                                                                                                                                                                                                                                                                                                      
   28 |     header = reinterpret_cast<const ElfEhdr *>(mapped);                                                                                                                                                                                                                                                                                                                                                                                                                             
```